### PR TITLE
Avoid mutable argument defaults outside of tests

### PR DESCRIPTION
**Pull request by hendrikmakait**
Removing mutable default arguments everywhere outside of tests to avoid footguns. Issues have been spotted using `flake8-bugbear`. 

XREF #6662
 
- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/http/routing.py [hunk 1]
```diff
@@ -7,7 +7,9 @@
 from tornado import web
 
 
-def _descend_routes(router, routers=set(), out=set()):
+def _descend_routes(router, routers=None, out=None):
```
Comment (hendrikmakait) [inline distributed/http/routing.py:10]: `test_sitemap` fails because of this change, so it seems like we're relying on an anti-pattern.

Comment (hendrikmakait) [reply distributed/http/routing.py:10]: Never mind, this relied on being able to pass in an empty set.

**Comment by fjetter** [comment]
I anticipate that https://github.com/dask/distributed/pull/6487 will pop up here as well (at least it should if bugbear is working properly)

cc @jacobtomlinson

**Comment by hendrikmakait** [comment]
> I anticipate that https://github.com/dask/distributed/pull/6487 will pop up here as well (at least it should if bugbear is working properly)

That's what I thought as well, but at the moment it looks like B006 only works for function/method arguments. See https://github.com/PyCQA/flake8-bugbear/issues/253

File: distributed/client.py [hunk 2]
```diff
@@ -4222,9 +4222,12 @@ def get_versions(self, check=False, packages=[]):
 
         >>> c.get_versions(packages=['sklearn', 'geopandas'])  # doctest: +SKIP
         """
-        return self.sync(self._get_versions, check=check, packages=packages)
+        return self.sync(self._get_versions, check=check, packages=packages or [])
 
-    async def _get_versions(self, check=False, packages=[]):
+    async def _get_versions(
+        self, check: bool = False, packages: Sequence[str] | None = None
+    ):
```
Comment (graingert) [inline distributed/client.py:4229]: missing a return type here

```
class VersionsDict(TypedDict):
    scheduler: dict[str, dict[str, Any]]
    workers: dict[str, dict[str, dict[str, Any]]]
    client: dict[str, dict[str, Any]]
```

```suggestion
    ) -> VersionsDict:
```

File: distributed/protocol/compression.py [hunk 3]
```diff
@@ -152,7 +152,7 @@ def maybe_compress(
     min_size=10_000,
     sample_size=10_000,
     nsamples=5,
-    compression=dask.config.get("distributed.comm.compression"),
```
Comment (graingert) [inline distributed/protocol/compression.py:155]: I think this is something that needs to be set globally before the process starts and so should be a default kwarg here

File: distributed/protocol/tests/test_protocol.py [hunk 4]
```diff
@@ -121,6 +121,27 @@ def test_maybe_compress(lib, compression):
         assert compressions[rc]["decompress"](rd) == payload
 
 
+@pytest.mark.parametrize(
```
Comment (hendrikmakait) [inline distributed/protocol/tests/test_protocol.py:124]: This currently breaks on `main` due to the way the `compression` default value is implemented on `maybe_compress`.

File: distributed/client.py [hunk 5]
```diff
@@ -117,9 +117,6 @@
     "pubsub": PubSubClientExtension,
 }
 
-# Placeholder used in the get_dataset function(s)
-NO_DEFAULT_PLACEHOLDER = "_no_default_"
```
Comment (hendrikmakait) [inline distributed/client.py:121]: Driveby: Replaced `NO_DEFAULT_PLACEHOLDER` with existing and already imported `no_default` from `distributed.utils`

File: distributed/protocol/tests/test_protocol.py (hunk 4 above)
Comment (gjoseph92) [reply distributed/protocol/tests/test_protocol.py:124]: It seems reasonable to me to make this behavior work as you have it here.

**Review by gjoseph92 — APPROVED** [review]
